crates/ra_hir_ty/src/infer.rs

```diff
@@ -363,14 +363,28 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
     }
 
     fn resolve_associated_type(&mut self, inner_ty: Ty, assoc_ty: Option<TypeAliasId>) -> Ty {
+        self.resolve_associated_type_with_params(inner_ty, assoc_ty, &[])
+    }
+
+    fn resolve_associated_type_with_params(
+        &mut self,
+        inner_ty: Ty,
+        assoc_ty: Option<TypeAliasId>,
+        params: &[Ty],
+    ) -> Ty {
         match assoc_ty {
             Some(res_assoc_ty) => {
                 let ty = self.table.new_type_var();
+                let mut builder = Substs::build_for_def(self.db, res_assoc_ty).push(inner_ty);
+                for ty in params {
+                    builder = builder.push(ty.clone());
+                }
```

Comment:
this is `builder.fill(params.iter().cloned())`, right?

Reply:
except yours will check that the number of provided types matches exactly... maybe we need a `fill_exact`

Reply:
I changed it to: 
```rust
let builder = Substs::build_for_def(self.db, res_assoc_ty)
    .push(inner_ty)
    .fill(params.iter().cloned());
```